### ISSUE TITLE
Generalise allele feture subsitution from toml files

### ISSUE THREAD

**Issue by manulera** [manulera/genestorian_data_refinement#19]
Do this after #17 so you have the genes in toml file as well.

* [x] In the script you have written to extract the data from the Dey lab, add a function that:
  * Takes as an input:
    *  A list of genotypes.
    * A toml file with allele features (genes, alleles, tags, etc.).
    * A string that will be used to replace the allele features.
  * Replaces the allele features defined in the toml file by the string you pass as third argument (see example below)
> * Remember to replace the longest names or synonyms first, so that `kanMx6` gets replaced before `kan` etc.
> * Remember to make everything lowercase when replacing strings.

The function may look like this:

```python
def replace_allele_features( toml_file, genotypes, word):
    code

genotypes = ['cls1-36 ase1-GFP:KanMx6']
# Replace all the allele features included in alleles.toml by the string ALLELE
genotypes2 = replace_allele_features('data/alleles.toml', genotypes, 'ALLELE')
#genotypes2: ['ALLELE ase1-GFP:KanMx6']

# Replace all the genes included in genes.toml by the string GENE
genotypes3 = replace_allele_features('data/alleles.toml', genotypes2, 'GENE')
#genotypes3: ['ALLELE GENE-GFP:KanMx6']
```

Then try the code for alleles and genes as in the example above

**Comment by anamika-yadav99** [comment]
https://github.com/manulera/genestorian_data_refinement/pull/20/commits/c4e8e399506b9ddbafb52af3960578fdd7d4b51b

**Comment by anamika-yadav99** [comment]
https://github.com/manulera/genestorian_data_refinement/pull/20/commits/f89111677c7bf0711a7f170db8a5db21cdbea6fb
![image](https://user-images.githubusercontent.com/54634163/175898326-8a7319c5-2b8c-4459-808c-81e3468c99bc.png)

**Comment by anamika-yadav99** [comment]
https://github.com/manulera/genestorian_data_refinement/pull/20/commits/42f3e074e45f919c15702950ec96fdef96d8f9ae
@manulera Can you have a look at this?

**Comment by manulera** [comment]
Hello @anamika-yadav99 good job, that looks much better.

There are still some changes to be made so that it does what it's meant to do, some of it was mentioned in the call today, you can try these two inputs that will give you errors. Remember not to use the `ref` field, since it contains a url sometimes:

```python
genotypes = ['SPAC1002.01'] # Gene ids are not replaced by GENE
genotypes = ['cls1-36 ase1-GFP:KanMx6', 'SPAC1002.01'] # Does not work for lists with len>1 (see indentation line 38)
```

Also, you are not addressing this:
> Remember to replace the longest names or synonyms first, so that kanMx6 gets replaced before kan etc.

Some comments to improve readability of the code:

* The function `feature_dict` is a bit hard to follow. Here some tips to improve it:
  * You don't need the function `feature_name`, it is the same as calling `f.keys()[0]` and you are reading the whole toml file again. You can add a comment on top of the line where you do this to explain why you do it like this (so far, there is only one key for the whole file, which is shared by all the features listed in the toml file).
  * Give variables meaningful names:
    * In `f` the keys are the "feature types" (alleles, tags, etc.), and the value is a dictionary of features of that type. You can name this variable `feature_types_dict`, for example.
    * `f[f.keys()[0]]` is a dictionary of features, so you can assign it to a variable `features_dict` to improve readability vs. `f[toml_key_name]`.
    * `toml_key` represents the key of a feature in the list, you can call it `feature_key`.
    * `names` is a single string value, so it should not be in plural, should be called `name`.
    * `synonym` is a list of strings, so it should be a plural
  * You don't need to call `keys()` to iterate over the keys of a dictionary, it's the default: `for key in dict` is the same as `for key in dict.keys()`
  * Use consistent syntax to assign values to a dictionary:
    * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L21
    * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L25
    * I would favour the one in line 25, I think it's cleaner.

* Comments to improve `replace_allele_features`:
  * The variable `genotype` (in the function and the script) is a list of genotypes, it should be plural, `genotypes`.
  * The variable `alleles` stores the genotypes with substituted allele features, name it accordingly.
  * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L33
    * Here you are iterating over genotypes, not alleles (in this first version we are not splitting the genotypes into alleles yet)
  * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L35
    * Give `a` a meaningfull name
  * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L36
    * This should never be empty.
    * After this, you have found your match, so you can stop the loop
  * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L38-L42
    * Make this loop more clear and shorter: `for match in matches`
  * https://github.com/manulera/genestorian_data_refinement/blob/42f3e074e45f919c15702950ec96fdef96d8f9ae/Lab_strains/dey_lab/generalised_function.py#L45-L46
    * This is wrong, it would append `genotype` (which is currently a list) to the list `alleles`, so you would end up with `[genotype,]` instead of `genotype` as a return value. These errors are easier to avoid when you name your variables which are lists as plurals! 😉.

* Remove unused dependencies

**Comment by manulera** [comment]
Added in #23
